engine/src/scheme.rs

```diff
@@ -1,29 +1,53 @@
 use crate::{
     ast::FilterAst,
     functions::Function,
-    lex::{complete, expect, span, take_while, LexErrorKind, LexResult, LexWith},
-    types::{GetType, Type},
+    lex::{complete, expect, skip_space, span, take_while, LexErrorKind, LexResult, LexWith},
+    types::{GetType, RhsValue, Type},
 };
 use failure::Fail;
 use fnv::FnvBuildHasher;
 use indexmap::map::{Entry, IndexMap};
-use serde::{Deserialize, Serialize, Serializer};
+use serde::{ser::SerializeSeq, Deserialize, Serialize, Serializer};
 use std::{
     cmp::{max, min},
     error::Error,
     fmt::{self, Debug, Display, Formatter},
+    iter::Iterator,
     ptr,
 };
 
-#[derive(PartialEq, Eq, Clone, Copy)]
+#[derive(Debug, PartialEq, Eq, Clone, Serialize)]
+#[serde(untagged)]
+pub enum FieldPathItem {
+    MapKey(String),
+}
+
+// special case for simply passing strings
+impl From<&str> for FieldPathItem {
+    fn from(name: &str) -> Self {
+        FieldPathItem::MapKey(name.to_string())
+    }
+}
+
+#[derive(PartialEq, Eq, Clone)]
 pub(crate) struct Field<'s> {
     scheme: &'s Scheme,
     index: usize,
+    pub path: Vec<FieldPathItem>,
```

Comment:
I wonder if this is really semantically part of the field and not a higher-level thing... I'd probably move it to unary operations or something like that instead.

Reply:
I wondered the same but found the `path` notion an interesting solution to get and set value. It could also be used from FFI. I think it should be kept for those usage but maybe not to represent the AST itself.
Would you be then more inclined to introduce those new AST nodes:
* `MapKeyExpr` for now
* `ArrayIndexExpr` when arrays are introduced
* `SliceExpr` if we ever supports Wireshark native slice operations
?

Reply:
I think a single `IndexExpr` with `lhs` + an `Index` enum should be fine.